### DISCUSSION TITLE
Importing python modules for custom kitten extension

### DISCUSSION THREAD

**Discussion by OhMesch** [kovidgoyal/kitty#4610] [Q&A] [answered]
Hello!

I'm in the process of writing a custom kitten extension which requires importing a handful of python modules which don't seem to already be included by Kitty.

Running Ubuntu 20.04 and on kitten launch am getting Python "ModuleNotFoundError"

Classic pip install gets the modules on my machine and working generally but clearly isn't the same version of python kitty is using. 
As far as I can tell kitty is using python 3.9.4, explicitly installing this version of python and using pip3.9.4 continues to error as before.

Eventually I found in the docs the following regarding python versioning:
"While kitty does use python, it is not a traditional python package, so please do not install it in site-packages. Instead run:
python3 setup.py linux-package"

How can I configure this nontraditional python with additional module requirements?

Any advice is greatly appreciated!

**Comment by kovidgoyal** [comment, marked as answer]
If you install the system kitty package it will use the system python.
If you install the kitty binary it uses its own python. If you want to
add pure python packages to the later, simply append the directory where
they are installed to sys.path before importing them.

**Reply by OhMesch** [reply]
Thank you very much!

**Comment by jamilraichouni** [comment]
Hi!

What is meant with '_If you install the system kitty package it will use the system python_.'?

---

**Edit:**
Okay, I think I get it now. On macos there is just the binary package. `kitty` on PyPI is something else.
So, `sys.insert(0, /path/to/additional/pkgs)` for a Python version that matches kitty's py version, right?

---

I'm quite interested in that.

I'm on macos and ran `pip3 install kitty` but when I completely close and restart `kitty`,
`import kitty` in a test kitten and set a breakpoint to ask for `kitty.__file__` it tells me

`'/Applications/kitty.app/Contents/Resources/Python/lib/kitty-extensions/python-lib.bypy.frozen/kitty/__init__.pyc'`

Also I find no additional kitty binary after that system-wide pip install for `kitty`.

Really great would be to be able to create some `~/.config/kitty/.venv`, pip install a `kitty` and any additional pkgs into it and run `kitty` from inside that venv. This is a bit what I read from that quoted sentence but it looks as if I misunderstood that!?

All best,
Jamil

**Comment by kovidgoyal** [comment]
kitty is not pip installable. Use a distro provided kitty package on
Linux, no clue on macOS.
